Extrinsic cal stubs

ethanrublee opened the pull request:
This adds apriltag, sophus, and ceres as dependencies, and a bit of a stub for the extrinsic calibration process, which just sends out a transform between tractor/base and the tracking_camera/front frame.

ethanrublee commented:
I may add apriltag detection to this branch.

isherman commented on cpp/farm_ng/extrinsic_calibrator.cpp:
```diff
@@ -0,0 +1,92 @@
+
+  void on_event(const farm_ng_proto::tractor::v1::Event& event) {
+    if (event.data().type_url() !=
+        "type.googleapis.com/farm_ng_proto.tractor.v1.NamedSE3Pose") {
```
could be something like `"type.googleapis.com/" + farm_ng_proto::tractor::v1::NamedSE3Pose.descriptor().full_name()`

ethanrublee replied:
thanks.  updated.

ethanrublee commented:
@isherman this currently runs detection and publishes it from the tracking_camera code.  Verified it runs on the tractor.

Suggest printing some april tags to test it out, hard coded for the tag36h11 group, you can find pngs here https://github.com/AprilRobotics/apriltag-imgs/tree/master/tag36h11

isherman commented:
> @isherman this currently runs detection and publishes it from the tracking_camera code. Verified it runs on the tractor.
> 
> Suggest printing some april tags to test it out, hard coded for the tag36h11 group, you can find pngs here https://github.com/AprilRobotics/apriltag-imgs/tree/master/tag36h11

sure, will make sure i can reproduce here before approving

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -5,14 +5,29 @@
 
+#include <glog/logging.h>
+
+#include <chrono>
```
was getting `error: 'mutex' in namespace 'std' does not name a type` until I added `#include <mutex>`

would https://include-what-you-use.org/ have caught this?

ethanrublee replied:
I'm in favor of adding iwyu to our pre-commits, and/or as a build step in cmake.  Let's do that in a future PR.

isherman commented on third_party/readme.md:
````diff
@@ -1,14 +1,51 @@
+sudo cp $FARM_NG_ROOT/third_party/librealsense/config/99-realsense-libusb.rules /etc/udev/rules.d/
 sudo udevadm control --reload-rules && udevadm trigger
 ```
+
````
I needed the following additional deps in `bootstrap.sh`:
```
libeigen3-dev \
libgoogle-glog-dev \
libopencv-dev \
```

ethanrublee replied:
Ok will add.

isherman commented on protos/farm_ng_proto/tractor/v1/apriltag.proto:
```diff
@@ -0,0 +1,47 @@
+
+    // The 3x3 homography matrix describing the projection from an
+    // "ideal" tag (with corners at (-1,1), (1,1), (1,-1), and (-1,
+    // -1)) to pixels in the image. This matrix will be freed by
```
Don't think the ownership note is relevant here

ethanrublee replied:
Thanks, sorry these comments are from apriltag library.

isherman commented on third_party/readme.md:
````diff
@@ -1,14 +1,51 @@
-```
-sudo cp $FARM_NG_ROOT/third_party/librealsense/config/99-realsense-libusb.rules /etc/udev/rules.d/ 
+```bash
+sudo cp $FARM_NG_ROOT/third_party/librealsense/config/99-realsense-libusb.rules /etc/udev/rules.d/
````
FYI, I've been running a single T265 on the Jetson successfully without this installation step

ethanrublee replied:
I think the debian install takes care of this for us.

ethanrublee replied:
Removing the submodule and these lines from the readme.

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -99,13 +114,18 @@ Event ToNamedPoseEvent(const rs2::pose_frame& rs_pose_frame) {
+
+  Sophus::SE3d se3;
+  ProtoToSophus(vodom_pose_t265.a_pose_b(), &se3);
+  SophusToProto(se3.inverse(), vodom_pose_t265.mutable_a_pose_b());
```
what's the thinking around this change?

ethanrublee replied:
At the moment, the visualization code only supports displaying poses which have their reference frames in a specific order.  Ideally the client would be smart enough to traverse the tree, and be able to apply the rule that a_pose_b == b_pose_a.inverse().

So yeah this is a hack.

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -114,19 +134,81 @@ Event ToNamedPoseEvent(const rs2::pose_frame& rs_pose_frame) {
+    tag36h11_destroy(tag_family_);
+  }
+
+  ApriltagDetections Detect(cv::Mat gray) {
```
better name than `gray`?

ethanrublee replied:
Need some image name conventions i guess.  gray is a pretty common name for arguments which expect a grayscale image.  Open to suggestions, suggest a follow up pass once we have a basic style guide.

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -114,19 +134,81 @@ Event ToNamedPoseEvent(const rs2::pose_frame& rs_pose_frame) {
+    zarray_t* detections = apriltag_detector_detect(tag_detector_, &im);
+    // Draw detection outlines
+    for (int i = 0; i < zarray_size(detections); i++) {
+      ApriltagDetection* detection = pb_out.add_detections();
```
nit: would move this two lines down for easier top-to-bottom readability (get the detection, then populate the proto)

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -114,19 +134,81 @@ Event ToNamedPoseEvent(const rs2::pose_frame& rs_pose_frame) {
+
+    ApriltagDetections pb_out;
+    zarray_t* detections = apriltag_detector_detect(tag_detector_, &im);
+    // Draw detection outlines
```
not actually drawing anything here, right?

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -188,10 +304,16 @@ class TrackingCameraClient {
   rs2::pipeline pipe_;
   std::mutex mtx_;
```
always appreciate a more descriptive name or, if not possible, comment to describe what's being protected

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -163,11 +260,30 @@ class TrackingCameraClient {
       if (count_ == 0) {
         writer_->write(frame_0);
+        if (!detection_in_progress_) {
+          detection_in_progress_ = true;
```
do we need to lock before changing this?

ethanrublee replied:
This is covered by the scope of the lock above.

isherman commented on protos/farm_ng_proto/tractor/v1/geometry.proto:
```diff
@@ -43,3 +43,9 @@ message NamedSE3Pose {
   string frame_b = 3;
 }
+
+message Matrix {
```
not sure `geometry.proto` is the right home for this long term

isherman commented:
confirmed apriltag detection working (saw an apriltag event with the correct tag ID when running `farm-ng-ipc-logger`)
 
definitely felt some of the pain of non-hermetic development environment reviewing this. i think we should make some investments in this area soon, since we have deps spread across readmes, bootstrap.sh, all with side effects on the local system. the name `bootstrap.sh` is misleading, if it needs to be run everytime an additional system dependency is required.

a link to a CPP style guide might be useful as the number of contributors grows. i would say the naming style so far trends towards short names & abbreviations more than my personal preference, but i'm sure you have strong opinions on that topic.

also minor note, struggled a bit to find a workflow that would let me suggest a set of local changes on top of this PR. since the branch is in your personal forked repo. i can't push to it. and couldn't find an alternative blessed github workflow to get that diff somewhere you could just 'accept changes'.

ethanrublee commented:
Thinking of using this style guide:
https://google.github.io/styleguide/cppguide.html

Would be helpful to have some hooks in pre-commit to check clang-format, iwyu, and any style nits.  But let's save for a little later - suggest we make the build/deps handling better, and then incorporate these items.  For now, we can review for google style readability.

ethanrublee commented:
Ok, i made a pass at addressing review comments Ian. Also, I added you as a contributor to my fork, so if you would like to push some changes, you can add me as a remote with push access.

```
git remote add ethanrublee git@github.com:ethanrublee/tractor.git
```

isherman commented:
> Ok, i made a pass at addressing review comments Ian. Also, I added you as a contributor to my fork, so if you would like to push some changes, you can add me as a remote with push access.
> 
> ```
> git remote add ethanrublee git@github.com:ethanrublee/tractor.git
> ```

Cool, I think this is easier for now than an additional fork

isherman commented on cpp/farm_ng/tracking_camera.cpp:
```diff
@@ -274,13 +286,15 @@ class TrackingCameraClient {
 
             event_bus_.Send(event);
-            std::lock_guard<std::mutex> lock(mtx_);
+	    // signal that we're done detecting, so can post another frame for detection.
```
some funky indentation in this file
